Check remote git URL if it is accessible

jscotka opened the pull request:
fixes: https://github.com/psss/tmt/issues/720

thrix commented on tmt/utils.py:
```diff
@@ -964,6 +964,10 @@ def public_git_url(url):
     """
 
+    # Using internal url instance transformtion is not supported
+    if re.match(r'ssh://.*osci\.redhat\.com/.*', url):
```
I do not like this approach :( hardcoding some specific URL here. Wouldn't be better to say we do not support git ssh backend at all? Or to support it?

jscotka replied:
I also don't like it, but theorerically for distgit it is safe to change `git` or `ssh` -> https

but also some public instances could have different path when using ssh, git, https, and probably whoever using ssh or git probably could have trouble to git clone without credentials.

thrix commented on tmt/utils.py:
```diff
@@ -964,6 +964,10 @@ def public_git_url(url):
+    if re.match(r'ssh://.*osci\.redhat\.com/.*', url):
+        raise ConvertError(f"Use distgit, transformation for {url} is not supported")
+
     # GitHub, GitLab
```
if this is about gitlab/github, we should rather match below gitlab/github, it is not tied to gitlab/github at all in the regex ....

jscotka replied:
At all this trasnformation probably should be just for page where we know it is possible to trasform as you said. and aovid transform any url from some protocol to `https`

But still not sure if there is good way how to do it. maybe when doing this transformation to check if the target is accessible.
If not raise error or use non-transformed url. But this check of external resources will cost some time.

lukaszachy replied:
We have #481 for external resources - since this happens sporadically it is better to play safe and make sure reference is correct

lukaszachy replied:
And with that said - maybe is this check obsolete? Because we need to assert whole fmf-id

jscotka replied:
@lukaszachy Theoretically it is obsolote in case it also checks who fmf-id, but in fact. it is closer to git and transformations, which are not sometimes correct and it is better to check.

But finally my solution helps to solve chicken-egg problem. because it is not good to commit changes to git, without calling tmt test export --nitrate --general before, because it modifies fmf files, and also it is not possible to export it firstly, because in case you'll try to check whole `fmf_id` proper data could not be commited to git, so that
and catch it early

At least for export it could be problem to check full fmf_id identifier.

For lint it is okay.

lukaszachy replied:
I beg to differ: It is easy to run tmt export but forgot to run git push and that is thing we have to catch.

jscotka replied:
@lukaszachy right.This PR is very simple and just close to transformation and find it early that final fmf_id url will be bad.
But in fact check full fmf identifier is superset of this simple check.

Yes, this commit&push miss is bad, but you cannot do that direclty for export, because it is expected to not have it commited and pushed to git in export phase, but just via some tmt test lint or what, but you can also forgot to call `fmf lint` so that it does not solve this issue.

For your issue is better to print some message at the end with **big red warning** do not forget to to commit and push changes to git, after export.

lukaszachy replied:
> ..., because it is expected to not have it commited ...

Not sure. While a bit annoying (two commits) we shouldn't lead user to create race in TCMS.

jscotka replied:
@lukaszachy exactly, so that for export you are not able to check full fmf_id check as it does not exists inside remote URL without commit and push, so this PR helps you to check at least `url` what is probably the worst case.

jscotka commented:
/packit test

psss reviewed (approved):
Seems better than nothing to me. I'm ok with merging after a couple of changes in 360837d unless @lukaszachy or @thrix disagree. Moving to `1.7` though, to give it some more time for discussion and at least a bit more testing before the fedora release as the original version broke `tmt test show -v`.

psss reviewed (changes requested):
Hm, copr build failed because of no network access. More changes will be necessary.

jscotka commented:
> Hm, copr build failed because of no network access. More changes will be necessary.

Yep, I've fixed the tests by fixture and simple mocking and also rebased

psss commented:
Would be better to add new commits during the review and [squash changes](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests) at the end. It's hard to see what actually has changed after three force pushes. Plus it seems some of my changes in 360837d have been lost (the `tmt test show -v` regression is back).

jscotka commented:
> Would be better to add new commits during the review and [squash changes](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests) at the end. It's hard to see what actually has changed after three force pushes. Plus it seems some of my changes in [360837d](https://github.com/psss/tmt/commit/360837d8bdca654dd0eb64efe4aadb07757500ae) have been lost (the `tmt test show -v` regression is back).

Yep, I'll use this way instead of implicit squashing when I found that it is not necessarry to see older commit. Understand that in process of review it is hard when I change it by force pushes.

But please let me show what I've exactly forgot when applying your commit. or better to say. I've called also pre-commit manually. so possible that precommit did something else, than you, but similar, but at lest test messages and comment were changed exactly to your suggestions.

and I also need to know what `tmt test show -v` regression is back means? I do not know what exactly you are talking about and what is bad there.

psss commented:
> But please let me show what I've exactly forgot when applying your commit.

The changes in the `check_git_url()` method [itself](https://github.com/psss/tmt/commit/360837d8bdca654dd0eb64efe4aadb07757500ae#diff-ccf2e34f34654a98eb09412ae8a65d9c8b1650c7697189603998f54a34e906b9R995).

> I also need to know what `tmt test show -v` regression is back means?

Try to run it ;-) The output from `git ls-remote`is not captured and is shown to the user.

jscotka commented:
> > But please let me show what I've exactly forgot when applying your commit.
> 
> The changes in the `check_git_url()` method [itself](https://github.com/psss/tmt/commit/360837d8bdca654dd0eb64efe4aadb07757500ae#diff-ccf2e34f34654a98eb09412ae8a65d9c8b1650c7697189603998f54a34e906b9R995).
> 
> > I also need to know what `tmt test show -v` regression is back means?
> 
> Try to run it ;-) The output from `git ls-remote` is not captured and is shown to the user.

@psss thank you very much after calling pre-commit I did not see this improvement and also as I run the mentioned command I did not see the output of the command, because output were too big that blind to see it :-)

psss commented on tmt/utils.py:
```diff
@@ -991,12 +991,10 @@ def create_file(
 def check_git_url(url):
```
Let's also include the docstring proposed in 360837d.

psss commented:
Sorry, for late response. We've shortly discussed this today on the hacking session and there are concerns that after this change it's not possible to list tests in verbose mode without a working internet connection. This does not play nicely with the expectation that a user should be able to clone a git repository and then be able to both implement and test offline without any external dependencies.

Furthermore, the current implementation fails if there is no git repository at all (probably not frequent but possible use case):

```
tmt init -t base
tmt test show -v
/tests/example
     summary Concise summary describing what the test does
        test ./test.sh
        path /tests/example
   framework beakerlib
      manual false
     require beakerlib
    duration 5m
     enabled true
      result respect
     sources /tmp/base/tests/example/main.fmf
Unable to contact remote git via ''.
```

But the current implementation is not much better as it produces a bunch of warnings:

```
/tests/example
     summary Concise summary describing what the test does
        test ./test.sh
        path /tests/example
   framework beakerlib
      manual false
     require beakerlib
    duration 5m
     enabled true
      result respect
     sources /tmp/base/tests/example/main.fmf
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
      fmf-id name: /tests/example
             url:
             ref:
             path: /.
```

Need to discuss this more and find the best way forward.

psss reviewed (changes requested):
Thanks for implementing this. Seems to be working fine for the new test case creation. I think it would make sense to include the `url` check during update of an existing test case as well. What do you think? Could you please add it?

lukaszachy commented:
>   I think it would make sense to include the url check during update of an existing test case as well. What do you think? Could you please add it?

Yes please - Right now I'm fixing rogue update of all cases with ref pointing to PR. This should have been prevented by this check.

psss commented:
> Yes please - Right now I'm fixing rogue update of all cases with ref pointing to PR. This should have been prevented by this check.

@jscotka, are you going to extend the check for update as well?

jscotka commented:
Hi @lukaszachy @psss sure, I can also add it for update of existing testcases.

psss commented:
Great, thanks! Shall we try to squeeze it into today's release?

psss reviewed (approved):
Thanks for all the updates. Looks good now. The `requre` data changes are quite large. I've applied `requre-patch purge` to zero newly introduce latency values. Squashed, ready for merge.
